### ISSUE TITLE
Proposal: a \\ b  becomes:  Math.ceil(a / b) to complement the existing // operator

### ISSUE THREAD

**Issue by blitmap** [jashkenas/coffeescript#4235]
Title says it all -

I'm not sure if this has been proposed already, I did make an effort to search this time. :smile:

``` coffeescript
a \\ b
```

would become:

``` javascript
Math.ceil(a / b)
```

Coffeescript already supports `//` to to the same thing with `Math.floor()`, I felt this was missing.

Any concerns?

**Comment by michaelficarra** [comment]
When would someone want to `Math.ceil(a / b)`?

**Comment by blitmap** [comment]
I think I was mostly surprised `\\` didn't exist but `//` did.

Recently I was writing Connect 4 (the game) in Coffeescript for a programming interview.  In that game I was using `//` to dynamically find the last index within a row by dividing the array holding everything by the hardcoded number of columns.  Using `//` gave me that last index for each row.

I could see `\\` being useful for situations like that to find the length of the row, rather than the last index.  To me it's very concise to use `//`, so I was hoping `\\` would also be available for those other situations.

The board in the game had a hardcoded number of columns and rows, so I could have easily gotten the last index and length of each row, but I defined the class methods of the Board to get this dynamically so I could feed it boards of an arbitrary [balanced] size and it wouldn't break in the future.  `//` came in handy a lot.

(Sorry for using an interview challenge as an exercise, I just see it being used for length vs last-index situations).

**Comment by dadleyy** [comment]
```
total_pages  = items.length \\ items_per_page
```

**Comment by blitmap** [comment]
@dadleyy I think you win the prize for the most concise and realistic example.

**Comment by michaelficarra** [comment]
I'm not convinced that a new operator is justified to avoid writing `+ 1`.

**Comment by blitmap** [comment]
I just saw it as an operator that was unlikely to be proposed for another use - to me it complements `//` with natural symmetry.

You know I tend to monkey-patch Array with:

```coffeescript
Array::last_index ?= -> @length - 1
```

I try to avoid `- 1` and `+ 1` when dealing with indexes or lengths because it makes the larger calculation harder to "skim".  I just thought this could be useful. :-)

**Comment by dadleyy** [comment]
I think the case for symmetry with `//` is enough argument alone; if `\\` is not used for `Math.ceil(a/b)`, what will it be used for? If it is used for something else, wouldn't that cause confusion? If it is not used for anything else, isn't that a waste of a safe extension in the language that would be beneficial to the users of the language?

**Comment by lydell** [comment]
@michaelficarra `+ 1` is _not_ the same:

```coffee
5 // 2 == 2
5 \\ 2 == 3
5 // 2 + 1 == 3 # ok

4 // 2 == 2
4 \\ 2 == 2
4 // 2 + 1 == 3 # oops, probably didn't want +1 here
ceiledDivision = (a, b) -> if a % b == 0 then a / b else a // b + 1 
ceiledDivision(4, 2) == 2
ceiledDivision2 = (a, b) -> Math.ceil a / b
ceiledDivision2(4, 2) == 2
```

**Comment by atg** [comment]
I don't think the existence of `a \\ b` would do harm, but if I were doing this operation a lot I would just write `ceil = Math.ceil` at the top of the file and use `ceil(a / b)`.

**Comment by michaelficarra** [comment]
I agree with @atg. Just because syntax is available does not make it a good idea to use it. Syntax isn't *free*. It comes at a cost of added complexity of the language. An unfamiliar reader is going to have to look up these symbolic shorthands instead of just reading the definition of `ceiledDivision`. Don't get me wrong; this is okay for very common/familiar operations like `+` and `*`. But I'm still not convinced the trade-off is worth it here.

**Comment by dadleyy** [comment]
> An unfamiliar reader is going to have to look up these symbolic shorthands instead of just reading the definition of `ceiledDivision`

This same argument could be used for the `//` operator... Is `Math.floor` in some way more important than `Math.ceil`?

**Comment by michaelficarra** [comment]
@dadleyy Slippery slope arguments aren't going to work well to convince me, as I'm already of the mindset that I'd much rather have no operators than more operators. Functions work just fine, but we're unfortunately still trying to incorporate maths notation into our programming languages for familiarity's sake. Also, yes, I believe floored division is much more common, as that's the default for integral division in many other languages.

**Comment by xixixao** [comment]
Backslashes are hard to deal with. Double backslashes are double hard to deal with. We should avoid them in source code. `Math.ceil` doesn't require a new operator (we also don't have `...x` to mean whatever just because `x...` means something). Integer division is a much more common operation.

**Comment by blitmap** [comment]
Backslashes (as I understand) are only valid within strings, heredocs, and regular expressions.  Outside of these contexts, it would *only* translate to `Math.ceil(a / b)`.  I do not believe this is "double hard to deal with".

**Comment by trylks** [comment]
I see no symmetry in those operators because to me the `//` is the normal division. When doing integer division in most languages stemming from the C family the result would be another integer, which would correspond to `//` in dynamic languages. The way to implement integer division when translating to JavaScript is to do Math.floor, but this is not about floor, ceil, or round (which would then be missing), it's integer division.

Therefore it's kind of traditional to have `/` and `//` in dynamic languages that may return float from integer division. On the contrary, to the best of my knowledge, `\\` would be new, weird, and slightly puzzling (I have not seen any other language with it). And when added, on that logic, we should consider a new operator for Math.round. It would therefore solve a symmetry problem that does not exist, and would create a new one where Math.round is missing. There are also divisions to infinity and minus infinity, that do not even seem to be part of the standard JavaScript library, but are in fact Math.floor and Math.ceiling. As far as I remember from my lessons (many years ago), floor of -6 / 5 should be -1, and division to -infinity (the current implementation of floor) is -2.

If you want to entirely forget about the Math library, you can still use integer division in any language that supports it (like CoffeeScript, abstracting from the way that the integer division is implemented, with the `//` operator). Like this:

```
floordiv = (a, b) -> a // b
rounddiv = (a, b) -> (2 * a + b) // (b * 2) # not the best for very large numbers, but I think it's elegant
ceildiv = (a, b) -> (a - 1 + b) // b
``` 

I just think that the integers deserve some love. Most probably, for practical purposes in the particular case of CoffeeScript, it is better to use the solution from @atg 

If still having Math.floor and not Math.ceiling seems inconsistent, I would propose to change Math.floor with Math.trunc and have the "symmetry" of not having neither Math.floor nor Math.ceiling.

This has the advantage of having more symmetry, and having -6 // 5 == -1 and 6 // 5 == 1. This kind of symmetry is lost in languages like Python, and while I think that all of them should change and be consistent with this, that is not going to happen.  If you think that I'm crazy, tell me how crazy is this: (-6//5)//(-(6//5)) == 2

In short: If we care about symmetry, I have a different symmetry that I think would be more relevant and better. My suggestion is to care more about convention and less about symmetry. Either we care about symmetry or convention, it is better to not add the proposed operator.

**Comment by GeoffreyBooth** [comment]
Closing as there doesn’t seem to be consensus that this should be added.
